### DISCUSSION TITLE
CS8602 despite guard call in generic method

### DISCUSSION THREAD

**Discussion by skrysm** [dotnet/csharplang#3954] [General] [answered]
Say I have this (simplified) code:

```c#
public static class Guard
{
    [ContractAnnotation("value:null => halt")]
    public static void IsNotNull<T>([ValidatedNotNull] T value, string paramName)
    {
        if (value is null)
        {
            throw new ArgumentNullException(paramName);
        }
    }
}

public static class SecureStringHelper
{
    public static string ToStringSecure<T>(T value)
    {
        Guard.IsNotNull(value, nameof(value));

        if (value is SecureString)
        {
            return "<masked>";
        }

        return value.ToString(); // gives: CS8602: Dereference of a possibly null reference.
    }
}
```

In the method `SecureStringHelper.ToStringSecure()` I'm getting a CS8602 (Dereference of a possibly null reference.) for `value.ToString()`.

This is because the compiler doesn't know that `Guard.IsNotNull(value, nameof(value));` has already verified that `value` is not null.

Is there a way to communicate this behavior to the C# compiler?

Other code flow analysis tools can already do this:

* For ReSharper, I provide `[ContractAnnotation("value:null => halt")]`
* For FxCop - specifically [CA1062](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1062) - I provide the `[ValidatedNotNull]` attribute.

I know I could write `value!.ToString()` but I'd rather have the compiler pick up on this automatically.

**Comment by 333fred** [comment, marked as answer]
Yes, via the `NotNull` attribute on the `value` parameter. See https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/nullable-analysis#specify-post-conditions-maybenull-and-notnull

**Reply by skrysm** [reply]
Just for anyone else looking for this: the `[NotNull]` attribute must be applied to the `value` parameter in `IsNotNull()`.
